Use score instead of likelihood in Vanilla Gradients

AlexKubiesa opened the pull request:
Addresses #159 

I've modified VanillaGradients.explain so instead of using the full model, it first creates a "score model" - the same as the original model, except without the final activation layer - and uses that instead.

If callers supply a model where the last layer isn't an activation layer, a warning is generated, and we use the original model as the score model. This tends to produce uninteresting output, so it should be pretty clear to users that something has gone wrong in this case.

Calling VanillaGradients.explain directly from VanillaGradientsCallback caused tf.function rerouting because the score model was being regenerated on every epoch. This made Tensorflow generate lots of warning messages and presumably reduces performance (I think I did see a performance hit when I tried it locally). So, instead, I split VanillaGradientsCallback.explain into "get_score_model" and "explain_score_model". The idea is that get_score_model gets called only once by the callback, and explain_score_model gets called every epoch. This is possible because the base Callback class has a set_model method which gets called by the framework, and I added an override for it.

- [x] Added code is tested
- [x] I've run Black and linter

RaphaelMeudec commented on tf_explain/core/vanilla_gradients.py:
```diff
@@ -1,13 +1,29 @@
+the model should be: a layer which computes class scores with no activation, \
+followed by an activation layer."
+
+_activation_layer_classes = (
+    tf.keras.layers.Activation,
+    tf.keras.layers.LeakyReLU,
+    tf.keras.layers.PReLU,
+    tf.keras.layers.ReLU,
+    tf.keras.layers.Softmax,
+    tf.keras.layers.ThresholdedReLU,
+)
```
Maybe you can check dynamically that the layer is in `tf.keras.activations`. So we don't have to store all the activation types on our side

AlexKubiesa replied:
I don't think that would work because the things in `tf.keras.activations` are activation functions rather than layers. Unfortunately there isn't really anything to distinguish an activation layer in tf.keras from any other type of layer...

But I assume the vast majority of models people will be applying Vanilla Gradients to will have softmax as the last activation function so as long as we have `Activation` and `Softmax` on this list, I think we will catch most cases.

RaphaelMeudec commented on tests/conftest.py:
```diff
@@ -55,3 +55,52 @@ def convolutional_model(random_data):
+
+
+@pytest.fixture()
+def convolutional_model_for_vanilla_gradients(random_data):
```
I think naming could be better with something that really explains what's the difference for vanilla gradients. Something like `convolutional_model_with_independant_activation_layer`

RaphaelMeudec commented on tests/conftest.py:
```diff
@@ -55,3 +55,52 @@ def convolutional_model(random_data):
+
+
+@pytest.fixture()
+def score_model_for_vanilla_gradients(random_data):
```
Same, maybe `convolutional_model_without_final_activation`

AlexKubiesa replied:
Yes, good point! I'll rename these.

RaphaelMeudec commented on tf_explain/core/vanilla_gradients.py:
```diff
@@ -24,15 +40,66 @@ def explain(self, validation_data, model, class_index):
             validation_data (Tuple[np.ndarray, Optional[np.ndarray]]): Validation data
                 to perform the method on. Tuple containing (x, y).
-            model (tf.keras.Model): tf.keras model to inspect
+            model (tf.keras.Model): tf.keras model to inspect. The last two layers of
+                the model should be: a layer which computes class scores with no
+                activation, followed by an activation layer.
```
I think it would be nice to explain why we have to do that

AlexKubiesa replied:
Ok, I'll add an explanation.

RaphaelMeudec commented on tf_explain/core/vanilla_gradients.py:
```diff
@@ -56,19 +123,13 @@ def compute_gradients(images, model, class_index):
-                expected_output, predictions
-            )
+            scores = model(inputs)
+            scores_for_class = scores[:, class_index : class_index + 1]
```
Why `scores[:, class_index : class_index + 1]` and not just `scores[:, class_index]` ?

RaphaelMeudec commented on tf_explain/core/vanilla_gradients.py:
```diff
@@ -1,13 +1,29 @@
 from tf_explain.utils.saver import save_grayscale
 
 
+UNSUPPORTED_ARCHITECTURE_WARNING = "\
+Unsupported model architecture for VanillaGradients. The last two layers of \
+the model should be: a layer which computes class scores with no activation, \
+followed by an activation layer."
```
It might be a personal opinion, but I'm not a fan of the line breakers. Could you format it this way instead?

```python
UNSUPPORTED_ARCHITECTURE_WARNING = (
    "Unsupported model architecture for VanillaGradients. The last two layers of "
    "the model should be: a layer which computes class scores with no activation, "
    "followed by an activation layer."
)
```

AlexKubiesa replied:
Ah, I didn't know you could do it that way!

RaphaelMeudec commented on tests/core/test_vanilla_gradients.py:
```diff
@@ -1,20 +1,47 @@
+
+def test_get_score_model_logs_warnings_when_model_not_suitable(convolutional_model):
+    explainer = VanillaGradients()
+    with pytest.warns(UserWarning, match=r"^Unsupported model architecture"):
```
Maybe you can import `UNSUPPORTED_ARCHITECTURE_WARNING` to make sure there is an exact match? It avoids test to fail if we change the formulation in the future.

AlexKubiesa replied:
Ok

RaphaelMeudec commented:
Thanks for the PR @AlexKubiesa! I think the way you've done it is nice, I just put some minor comments on the PR. Thanks again!

RaphaelMeudec commented:
@AlexKubiesa could you rebase your PR on master? I've fixed the failing tests, and migrated from Travis to Github Actions. Thanks!

AlexKubiesa commented:
Thanks for having a look @RaphaelMeudec, I'll push again once I've addressed all the comments.

By the way, it looks like the tox.ini file is gone from master, which means tox doesn't run for me locally after I rebased. Is that intended?

AlexKubiesa commented:
I've rebased now and I think I've addressed all the comments.

RaphaelMeudec commented:
All good! Thanks for the PR @AlexKubiesa!
